### PR TITLE
Uncurried versions of 'bracket' and 'bracketExit'

### PR DESCRIPTION
**Pull request by LGLO**
I open this to have version of `bracket` suitable for hot-path in performance sensitive applications.

### CODE REVIEW

**Comment by jdegoes** [comment]
@LGLO Thanks a lot for doing this! Let's merge after the build is passing.

**Comment by LGLO** [comment]
👍 I'll try to retain scaladoc link, although it's tricky.
When I do scaladoc fix, then I'll also add specs for inference (to see that besides compilation there are no cast exceptions).

**Comment by jdegoes** [comment]
@LGLO Would love to get this in as soon as the build passes!

**Comment by LGLO** [comment]
@jdegoes I've run this benchmark:
```
  @Benchmark
  def curriedBracket: UIO[Unit] = {
      ZIO.bracket(ZIO.unit)(_ => ZIO.unit)(_ => ZIO.unit)
  }
  @Benchmark
  def tupledBracket: UIO[Unit] = {
    ZIO.bracket(ZIO.unit, (_: Unit) => ZIO.unit, (_: Unit) => ZIO.unit)
  }
```
5 iterations, 30 seconds each (to let GC work) and observed no difference:
```
[info] Benchmark                                                                        Mode  Cnt     Score     Error   Units
[info] BracketBenchmark.curriedBracket                                                  avgt    5    28.820 ±   3.596   ns/op
[info] BracketBenchmark.curriedBracket:·gc.alloc.rate                                   avgt    5  4429.676 ± 552.787  MB/sec
[info] BracketBenchmark.curriedBracket:·gc.alloc.rate.norm                              avgt    5   136.002 ±   0.014    B/op
[info] BracketBenchmark.curriedBracket:·gc.churn.CodeHeap_'non-profiled_nmethods'       avgt    5     0.008 ±   0.003  MB/sec
[info] BracketBenchmark.curriedBracket:·gc.churn.CodeHeap_'non-profiled_nmethods'.norm  avgt    5    ≈ 10⁻⁴              B/op
[info] BracketBenchmark.curriedBracket:·gc.churn.G1_Old_Gen                             avgt    5  4437.325 ± 550.946  MB/sec
[info] BracketBenchmark.curriedBracket:·gc.churn.G1_Old_Gen.norm                        avgt    5   136.237 ±   0.566    B/op
[info] BracketBenchmark.curriedBracket:·gc.count                                        avgt    5  1434.000            counts
[info] BracketBenchmark.curriedBracket:·gc.time                                         avgt    5  1306.000                ms
[info] BracketBenchmark.tupledBracket                                                   avgt    5    28.790 ±   3.233   ns/op
[info] BracketBenchmark.tupledBracket:·gc.alloc.rate                                    avgt    5  4433.757 ± 496.258  MB/sec
[info] BracketBenchmark.tupledBracket:·gc.alloc.rate.norm                               avgt    5   136.000 ±   0.001    B/op
[info] BracketBenchmark.tupledBracket:·gc.churn.CodeHeap_'non-profiled_nmethods'        avgt    5     0.009 ±   0.003  MB/sec
[info] BracketBenchmark.tupledBracket:·gc.churn.CodeHeap_'non-profiled_nmethods'.norm   avgt    5    ≈ 10⁻⁴              B/op
[info] BracketBenchmark.tupledBracket:·gc.churn.G1_Old_Gen                              avgt    5  4443.315 ± 493.961  MB/sec
[info] BracketBenchmark.tupledBracket:·gc.churn.G1_Old_Gen.norm                         avgt    5   136.294 ±   0.773    B/op
[info] BracketBenchmark.tupledBracket:·gc.count                                         avgt    5  1558.000            counts
[info] BracketBenchmark.tupledBracket:·gc.time                                          avgt    5  1356.000                ms
```
I cannot prove this PR helps in anything.
I guess it should be closed.

**Comment by jdegoes** [comment]
@LGLO I am ok merging, actually, because in the best case the JIT can eliminate short-lived allocations, but that depends a lot on the code being run, version of JRE, etc. At least with these versions, we have a guarantee of minimal allocation. What do you think?

File: core/jvm/src/test/scala/scalaz/zio/IOSpec.scala [hunk 1]
```diff
@@ -237,4 +240,81 @@ class IOSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRuntim
...
+        result <- ZIO.bracketExit(
+                   IO.succeed(42),
+                   (_: Int, _: Exit[_, _]) => release.set(true),
+                   (_: Int) => IO.succeed(0L)
+                 )
+        released <- release.get
+      } yield (result must_=== 0L) and (released must_=== true)
+    }
+
+  object UncurriedBracketCompilesRegardlessOrderOfEAndRTypes {
```
Comment (LGLO) [inline core/jvm/src/test/scala/scalaz/zio/IOSpec.scala]: I don't really know where to place these `if-compiles-works` tests.

File: core/shared/src/main/scala/scalaz/zio/ZIO.scala [hunk 2]
```diff
@@ -408,14 +408,25 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
+
   /**
    * Returns an effect that, if this effect _starts_ execution, then the
    * specified `finalizer` is guaranteed to begin execution, whether this effect
    * succeeds, fails, or is interrupted.
    *
    * Finalizers offer very powerful guarantees, but they are low-level, and
    * should generally not be used for releasing resources. For higher-level
-   * logic built on `ensuring`, see [[ZIO#bracket]].
+   * logic built on `ensuring`, see `ZIO#bracket`.
```
Comment (LGLO) [inline core/shared/src/main/scala/scalaz/zio/ZIO.scala]: I'm sorry for this one, but scaladoc finds two `bracket` methods now and I cannot find method when signature contains type bounded by `type LowerR` declared inside `ZIO`.

**Comment by LGLO** [comment]
Okey!
I've fixed doc and added test.

**Pull request merged**